### PR TITLE
Sign chunks of containers

### PR DESCRIPTION
**Pull request by midnightercz**

### CODE REVIEW

File: pubtools/_quay/iib_operations.py [hunk 1]
```diff
@@ -136,6 +136,7 @@ def _remove_index_image_signatures(
         current_signatures (list): List of current signatures.
         target_settings (dict): Target settings.
     """
+    set_aws_kms_environment_variables(target_settings, "cosign_signer")
```
Comment (querti) [inline pubtools/_quay/iib_operations.py:139]: Is this a part of the PR?

File: pubtools/_quay/signer_wrapper.py [hunk 2]
```diff
@@ -126,29 +136,36 @@ def sign_container(
...
             task_id (str): Task ID to identify the signing task if needed.
         """
-        LOG.debug(
-            "Signing container %s %s %s",
-            sign_entry.reference,
-            sign_entry.digest,
-            sign_entry.signing_key,
-        )
+        for sign_entry in sign_entries:
+            if not sign_entry:
```
Comment (querti) [inline pubtools/_quay/signer_wrapper.py:140]: What does this mean?

File: pubtools/_quay/signer_wrapper.py [hunk 3]
```diff
@@ -126,29 +136,36 @@ def sign_container(
             sign_entry (SignEntry): SignEntry to sign.
```
Comment (querti) [inline pubtools/_quay/signer_wrapper.py:136]: Update docstring?

File: pubtools/_quay/signer_wrapper.py [hunk 4]
```diff
@@ -126,29 +136,36 @@ def sign_container(
...
-            "Signing container %s %s %s",
-            sign_entry.reference,
-            sign_entry.digest,
-            sign_entry.signing_key,
-        )
+        for sign_entry in sign_entries:
+            if not sign_entry:
+                break
+            LOG.debug(
+                "Signing containers %s %s %s",
```
Comment (querti) [inline pubtools/_quay/signer_wrapper.py:143]: Should it be containers plural?

File: pubtools/_quay/signer_wrapper.py [hunk 5]
```diff
@@ -164,7 +181,11 @@ def _filter_to_sign(self, to_sign_entries: List[SignEntry]) -> List[SignEntry]:
...
     @log_step("Sign container images")
     def sign_containers(
-        self, to_sign_entries: List[SignEntry], task_id: Optional[str] = None, parallelism: int = 10
+        self,
+        to_sign_entries: List[SignEntry],
+        task_id: Optional[str] = None,
+        parallelism: int = 10,
+        chunk_size: int = 10,
     ) -> None:
         """Sign signing entries.
```
Comment (querti) [inline pubtools/_quay/signer_wrapper.py:191]: Update docstring?

File: pubtools/_quay/signer_wrapper.py [hunk 6]
```diff
@@ -176,11 +197,16 @@ def sign_containers(
...
-                self.sign_container,
+                self.sign_container_chunk,
                 [
                     FData(args=x)
                     for x in zip(
-                        to_sign_entries,
+                        flatten(
+                            [
+                                list(grouper(group, chunk_size))
+                                for k, group in groupby(to_sign_entries, key=lambda x: x.repo)
```
Comment (querti) [inline pubtools/_quay/signer_wrapper.py:207]: Why is it grouped by repo? Also, this code is almost incomprehensible. Can't it be written in a more understandable way?

**Review by emilyzheng — APPROVED** [review]

File: pubtools/_quay/tag_docker.py [hunk 7]
```diff
@@ -564,7 +564,11 @@ def copy_tag_sign_images(self, push_item: Any, tag: str, executor: Executor) ->
                     if outdated_manifests:
                         signer.remove_signatures(outdated_manifests, _exclude=current_signatures)
                     if SIGNER_BY_LABEL[_signer["label"]].pre_push:
-                        signer.sign_containers(to_sign_entries, task_id=self.task_id)
+                        signer.sign_containers(
+                            to_sign_entries,
+                            task_id=self.task_id,
+                            chunk_size=self.target_settings.get("signing_chunk_size", 10),
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply pubtools/_quay/tag_docker.py]: updated

**Pull request merged**